bokehjs/src/coffee/renderer/annotation/box_annotation.coffee

```diff
@@ -0,0 +1,78 @@
+_ = require "underscore"
+HasParent = require "../../common/has_parent"
+PlotWidget = require "../../common/plot_widget"
+properties = require "../../common/properties"
+
+class BoxAnnotationView extends PlotWidget
+  initialize: (options) ->
+    super(options)
+    @fill_props = new properties.Fill({obj: @model, prefix: ''})
+    @line_props = new properties.Line({obj: @model, prefix: ''})
+
+  render: () ->
+    ctx = @plot_view.canvas_view.ctx
+    [sleft, sright, stop, sbottom] = @_calc_dims()
+
+    ctx.save()
+
+    ctx.beginPath()
+    ctx.rect(sleft, stop, sright-sleft, sbottom-stop)
+
+    @fill_props.set_value(ctx)
+    ctx.fill()
+
+    @line_props.set_value(ctx)
+    ctx.stroke()
+
+    ctx.restore()
+
+  _calc_dims: () ->
+    canvas = @plot_model.get('canvas')
+    frame = @plot_model.get('frame')
+
+    sleft = @mget('left') ? frame.get('x_range').get('start')
+    sright = @mget('right') ? frame.get('x_range').get('end')
+    sbottom = @mget('bottom') ? frame.get('y_range').get('start')
+    stop = @mget('top') ? frame.get('y_range').get('end')
+
+    # refactor to pass x/y mapper names
+    [[sleft, sright], [stop, sbottom]] = frame.map_to_screen([sleft, sright],
+                                                             [stop, sbottom],
+                                                             canvas)
+    # ugly control logic, may refactor
+    if @mget('left_units') == 'screen' and @mget('left')?
+      sleft = canvas.vx_to_sx(@mget('left'))
+    if @mget('right_units') == 'screen' and @mget('right')?
+      sright = canvas.vx_to_sx(@mget('right'))
+    if @mget('bottom_units') == 'screen' and @mget('bottom')?
+      sbottom = canvas.vy_to_sy(@mget('bottom'))
+    if @mget('top_units') == 'screen' and @mget('top')?
+      stop = canvas.vy_to_sy(@mget('top'))
```

Comment:
We can talk about this tomorrow. I think for a given side (left, say), it actually is like this:
```
if @mget('left')?
    if @mget('left_units') == 'data'
        vleft = xmapper.map_to_screen(left)
    else 
        vleft = left
else
    vleft = frame.get('h_range').get('start')
sleft = canvas.vx_to_sx(vleft)
```
So this could maybe be abstracted into a function that takes the value, a mapper, and the frame range start/end:
```
sleft   = _compute_thing(@mget('left'),   xmapper, frame.get('h_range').get('start'))
sright  = _compute_thing(@mget('right'),  xmapper, frame.get('h_range').get('end'))
stop    = _compute_thing(@mget('top'),    ymapper, frame.get('v_range').get('end'))
sbottom = _compute_thing(@mget('bottom'), ymapper, frame.get('v_range').get('start'))
```